packages/medusa-plugin-webhooks/src/services/webhook.ts

```diff
@@ -0,0 +1,335 @@
+
+    return subscriptions;
+  }
+
+  public async send(subscription: BaseWebhook, payload: any): Promise<WebhookSendResponse> {
+    const { event_type, target_url } = subscription;
+
+    try {
+      const response = await fetch(target_url, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(payload),
+      });
+
+      const contentType = response.headers.get('content-type');
+      const data = contentType?.includes('json') ? await response.json() : await response.text();
+
+      return {
+        event_type,
+        target_url,
+        result: 'success',
+        data,
+      };
+    } catch (err) {
+      return this.onSendError(err, subscription, payload);
+    }
+  }
+
+  public async sendWebhooksEvents(webhooks: Webhook[], payload: any) {
+    const results = (await Promise.allSettled(
+      webhooks.map((webhook) => this.send(webhook, payload)),
+    )) as PromiseFulfilledResult<WebhookSendResponse>[];
+
+    results.forEach((result) => {
+      const resultMessage = result.value?.result === 'error' ? 'failed' : 'succeeded';
+
+      this.logger_.info(`Webhook ${result.value?.event_type} -> ${result.value?.target_url} ${resultMessage}.`);
+    });
+
+    return results;
+  }
+
+  public async createWebhookSubscription(data: Partial<Webhook>) {
+    return await this.atomicPhase_(async (transactionManager) => {
+      const webhookRepo = transactionManager.withRepository(this.webhookRepository_);
+
+      const subscription = webhookRepo.create(data);
+
+      await webhookRepo.save(subscription);
+
+      return subscription;
+    });
+  }
+
+  public async updateWebhookSubscription(id: string, data: any) {
+    return await this.atomicPhase_(async (transactionManager) => {
+      const webhookRepo = transactionManager.withRepository(this.webhookRepository_);
+
+      const existingWebhook = await webhookRepo.findOne({ where: { id } });
+
+      if (!existingWebhook) {
+        throw new MedusaError(MedusaError.Types.NOT_FOUND, `Webhook subscription with id: ${id} was not found.`);
+      }
+
+      const updatedWebhook = webhookRepo.merge(existingWebhook, data);
+
+      return await webhookRepo.save(updatedWebhook);
+    });
+  }
+
+  public async deleteWebhookSubscription(id: string) {
+    return await this.atomicPhase_(async (transactionManager) => {
+      const webhookRepo = transactionManager.withRepository(this.webhookRepository_);
+
+      const subscription = await webhookRepo.findOne({ where: { id } });
+
+      if (!subscription) {
+        throw new MedusaError(MedusaError.Types.NOT_FOUND, `Webhook subscription with id: ${id} was not found.`);
+      }
+
+      await webhookRepo.remove(subscription);
+
+      return subscription;
+    });
+  }
+
+  public async retrieveWebhooksOrderWithTotals(orderId: string) {
+    const blacklistedRelations = ['claims', 'swaps', 'children', 'returns', 'payments', 'refunds', 'region'];
```

Comment:
I'm wondering if for this plugin, we should make all the webhooks custom subscriptions... I don't love adding markethaus specific business logic on the default ones either... Thoughts?

Reply:
@cyberdude Could we have a way to override the defaults using a custom subscription?